### PR TITLE
Port swrad

### PR DESCRIPTION
**Pull request by andrewpauling**
Full SW scheme now validates

### CODE REVIEW

**Comment by andrewpauling** [comment]
I can't seem to get rid of the commits it is showing from older PRs that are already merged. Please ignore everything before "`Merge branch 'port_sw_firstloop' into port_swrad`"

File: Dockerfile [hunk 1]
```diff
@@ -67,7 +67,7 @@ RUN cd /serialbox && \
 
 # install gt4py
 RUN cd /
-RUN git clone -b tmp_forloop_fixes https://github.com/VulcanClimateModeling/gt4py.git
+RUN git clone -b v35 https://github.com/VulcanClimateModeling/gt4py.git
```
Comment (rheacangeo) [inline Dockerfile:70]: wahoo!

**Review by rheacangeo — COMMENTED** [review]
Sometimes the diff can get hosed up. Would you mind letting me know which files I should be looking at for this PR? Or maybe try starting a new branch and a new PR with the changes needed? I don't know what I should be looking at for this one.

**Comment by andrewpauling** [comment]
> Sometimes the diff can get hosed up. Would you mind letting me know which files I should be looking at for this PR? 

The files to look at are `swrad_gt4py.py`, stencils_gt4py.py` and the changes to the Dockerfile, which you already noted

> Or maybe try starting a new branch and a new PR with the changes needed? I don't know what I should be looking at for this one.

Sounds good, but I'm not sure how to do this.

**Review by rheacangeo — COMMENTED** [review]
It is really wild how much code this is! You mentioned the unit test vs whole test in the shortwave scheme, maybe that is meant as a future PR, not for this one. EDIT: now I see the 'do_test' section, testing one piece at a time after the full test. that works!

File: radiation/python/radsw/swrad_gt4py.py [hunk 2]
```diff
@@ -0,0 +1,1751 @@
...
+backend = "gtc:gt:cpu_ifirst"
+
+# Flag to say whether or not to do intermediate tests
+do_test = False
+
+ddir = "../../fortran/data/SW"
+ddir2 = "../../fortran/radsw/dump"
+
+serializer = ser.Serializer(ser.OpenModeKind.Read, ddir, "Generator_rank1")
+serializer2 = ser.Serializer(ser.OpenModeKind.Read, ddir2, "Serialized_rank1")
```
Comment (rheacangeo) [inline radiation/python/radsw/swrad_gt4py.py:29]: why are these in different directories? is this data from the full fortran model vs the standalone?

File: radiation/python/radsw/swrad_gt4py.py [hunk 3]
```diff
@@ -0,0 +1,1751 @@
...
+        )
+    else:
+        if var != "htswb":
+            valdict_final[var] = serializer.read(
+                var, serializer.savepoint["swrad-out-000000"]
+            )
+
+compare_data(outdict_final, valdict_final)
+
+if do_test:
```
Comment (rheacangeo) [inline radiation/python/radsw/swrad_gt4py.py:1629]: we might want to name this 'do_subtests' or 'unit_tests' (though they aren't really unit tests, since this is already doing a test.

File: radiation/python/radsw/cldprop_gt4py.py [hunk 4]
```diff
@@ -757,6 +764,6 @@ def cldprop(
     valdict[var] = serializer.read(
         var, serializer.savepoint["swrad-cldprop-output-000000"]
     )
-    outdict_np[var] = outdict_gt4py[var].view(np.ndarray).squeeze()
+    outdict_np[var] = outdict_gt4py[var][:, :, 1:, ...].view(np.ndarray).squeeze()
```
Comment (rheacangeo) [inline radiation/python/radsw/cldprop_gt4py.py:767]: was this a buffering issue?

File: radiation/python/radsw/setcoef_gt4py.py [hunk 5]
```diff
@@ -69,33 +70,33 @@
...
-outdict_gt4py["laytrop"] = create_storage_zeros(backend, shape_nlay, DTYPE_BOOL)
+outdict_gt4py["laytrop"] = create_storage_zeros(backend, shape_nlp1, DTYPE_BOOL)
 
 locvars = ["plog", "fp", "fp1", "ft", "ft1", "tem1", "tem2", "jp1"]
 
 locdict_gt4py = dict()
 for var in locvars:
     if var == "jp1":
-        locdict_gt4py[var] = create_storage_zeros(backend, shape_nlay, DTYPE_INT)
+        locdict_gt4py[var] = create_storage_zeros(backend, shape_nlp1, DTYPE_INT)
```
Comment (rheacangeo) [inline radiation/python/radsw/setcoef_gt4py.py:84]: why the change in indices?

File: radiation/python/radsw/stencils_sw_gt4py.py [hunk 6]
```diff
@@ -0,0 +1,4107 @@
...
+            # Prepare surface albedo: bm,df - dir,dif; 1,2 - nir,uvv.
+            albbm[0, 0, 0][0] = sfcalb[0, 0, 0][0]
+            albdf[0, 0, 0][0] = sfcalb[0, 0, 0][1]
+            albbm[0, 0, 0][1] = sfcalb[0, 0, 0][2]
+            albdf[0, 0, 0][1] = sfcalb[0, 0, 0][3]
+
+            zcf0 = 1.0
+            zcf1 = 1.0
+
+    with computation(FORWARD), interval(1, None):
```
Comment (rheacangeo) [inline radiation/python/radsw/stencils_sw_gt4py.py:180]: can these two computations be combined, using two separate 'with intervals'?

File: radiation/python/radsw/stencils_sw_gt4py.py [hunk 7]
```diff
@@ -0,0 +1,4107 @@
...
+    # Only the iovrlw = 1 option is ported from Fortran
+    with computation(PARALLEL), interval(2, None):
+        tem1 = 1.0 - cldf[0, 0, -1]
+
+        for n in range(ngptsw):
+            if cdfunc[0, 0, -1][n] > tem1:
+                cdfunc[0, 0, 0][n] = cdfunc[0, 0, -1][n]
+            else:
+                cdfunc[0, 0, 0][n] = cdfunc[0, 0, 0][n] * tem1
+
```
Comment (rheacangeo) [inline radiation/python/radsw/stencils_sw_gt4py.py:698]: I think this probably shouldn't be a PARALLEL computation, because it is both reading cdfunc[0,0,-1] and writing to cdfunc[0,0,0]. The validation may have gotten lucky, but with the vertical axis run in parallel you aren't guarenteel the layer at [0,0,-1] hasn't been changed before [0,0,0] is set.

File: radiation/python/radsw/firstloop_gt4py.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+backend = "gtc:gt:cpu_ifirst"
+
+sys.path.append(SERIALBOX_DIR + "/python")
+import serialbox as ser
+
+ddir = "../../fortran/data/SW"
+ddir2 = "../../fortran/radsw/dump"
+
+serializer = ser.Serializer(ser.OpenModeKind.Read, ddir, "Generator_rank1")
+serializer2 = ser.Serializer(ser.OpenModeKind.Read, ddir2, "Serialized_rank1")
```
[the remark this answers is not in the document]
Comment (andrewpauling) [reply radiation/python/radsw/firstloop_gt4py.py]: As discussed, this was for initial development and testing, the combined stencils are tested on all ranks

File: radiation/python/radsw/firstloop_gt4py.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+                zcf0 = zcf0 * (1.0 - cfrac)
+            elif iovrsw == 1:
+                if cfrac > ftiny:  # cloudy layer
+                    zcf1 = min(zcf1, 1.0 - cfrac)
+                elif zcf1 < 1.0:  # clear layer
+                    zcf0 = zcf0 * zcf1
+                    zcf1 = 1.0
+
+            elif iovrsw >= 2:
+                zcf0 = min(zcf0, 1.0 - cfrac)  # used only as clear/cloudy indicator
```
[the remark this answers is not in the document]
Comment (andrewpauling) [reply radiation/python/radsw/firstloop_gt4py.py]: It seems to be fine as long as they are not then read from in the same computation

File: radiation/python/radsw/swrad_gt4py.py (hunk 2 above)
Comment (andrewpauling) [reply radiation/python/radsw/swrad_gt4py.py:29]: Yes, `fortran/data/SW` is the data from fv3gfs-fortran, and the `dump` directory is for the standalone

File: radiation/python/radsw/swrad_gt4py.py (hunk 3 above)
Comment (andrewpauling) [reply radiation/python/radsw/swrad_gt4py.py:1629]: will update

File: radiation/python/radsw/cldprop_gt4py.py (hunk 4 above)
Comment (andrewpauling) [reply radiation/python/radsw/cldprop_gt4py.py:767]: Yes, I ended up making all fields size `nlp1` in the K dimensions for simplicity

File: radiation/python/radsw/setcoef_gt4py.py (hunk 5 above)
Comment (andrewpauling) [reply radiation/python/radsw/setcoef_gt4py.py:84]: Same as above

File: radiation/python/radsw/stencils_sw_gt4py.py (hunk 6 above)
Comment (andrewpauling) [reply radiation/python/radsw/stencils_sw_gt4py.py:180]: I think so, will do it

File: radiation/python/radsw/stencils_sw_gt4py.py (hunk 7 above)
Comment (andrewpauling) [reply radiation/python/radsw/stencils_sw_gt4py.py:698]: Yes, I've run into this elsewhere. I will change it
